File: packages/health_connector_hc_android/android/src/main/kotlin/com/phamtunglam/health_connector_hc_android/mappers/metadata_mappers/DeviceTypeMapper.kt
package com.phamtunglam.health_connector_hc_android.mappers.metadata_mappers

import androidx.health.connect.client.records.metadata.Device
import com.phamtunglam.health_connector_hc_android.pigeon.DeviceTypeDto

/**
 * Converts a [DeviceTypeDto] to a Health Connect device type constant.
 */
internal fun DeviceTypeDto.toHealthConnect(): Int = when (this) {
    DeviceTypeDto.UNKNOWN -> Device.TYPE_UNKNOWN
    DeviceTypeDto.WATCH -> Device.TYPE_WATCH
    DeviceTypeDto.PHONE -> Device.TYPE_PHONE
    DeviceTypeDto.SCALE -> Device.TYPE_SCALE
    DeviceTypeDto.RING -> Device.TYPE_RING
    DeviceTypeDto.FITNESS_BAND -> Device.TYPE_FITNESS_BAND
    DeviceTypeDto.CHEST_STRAP -> Device.TYPE_CHEST_STRAP
    DeviceTypeDto.HEAD_MOUNTED -> Device.TYPE_HEAD_MOUNTED
    DeviceTypeDto.SMART_DISPLAY -> Device.TYPE_SMART_DISPLAY
}

/**
 * Converts a Health Connect device type constant to a [DeviceTypeDto].
 */
internal fun Int.toDeviceTypeDto(): DeviceTypeDto = when (this) {
    Device.TYPE_UNKNOWN -> DeviceTypeDto.UNKNOWN
    Device.TYPE_WATCH -> DeviceTypeDto.WATCH
    Device.TYPE_PHONE -> DeviceTypeDto.PHONE
    Device.TYPE_SCALE -> DeviceTypeDto.SCALE
    Device.TYPE_RING -> DeviceTypeDto.RING
    Device.TYPE_FITNESS_BAND -> DeviceTypeDto.FITNESS_BAND
    Device.TYPE_CHEST_STRAP -> DeviceTypeDto.CHEST_STRAP
    Device.TYPE_HEAD_MOUNTED -> DeviceTypeDto.HEAD_MOUNTED
    Device.TYPE_SMART_DISPLAY -> DeviceTypeDto.SMART_DISPLAY
    else -> DeviceTypeDto.UNKNOWN
}

File: packages/health_connector_hc_android/lib/src/mappers/metadata_mappers/device_type_mapper.dart
import 'package:health_connector_core/health_connector_core_internal.dart'
    show DeviceType, sinceV1_0_0;
import 'package:health_connector_hc_android/src/pigeon/health_connector_hc_android_api.g.dart'
    show DeviceTypeDto;
import 'package:meta/meta.dart' show internal;

/// Converts [DeviceType] to [DeviceTypeDto].
@sinceV1_0_0
@internal
extension DeviceTypeDtoMapper on DeviceType {
  DeviceTypeDto toDto() {
    switch (this) {
      case DeviceType.unknown:
        return DeviceTypeDto.unknown;
      case DeviceType.watch:
        return DeviceTypeDto.watch;
      case DeviceType.phone:
        return DeviceTypeDto.phone;
      case DeviceType.scale:
        return DeviceTypeDto.scale;
      case DeviceType.ring:
        return DeviceTypeDto.ring;
      case DeviceType.fitnessBand:
        return DeviceTypeDto.fitnessBand;
      case DeviceType.chestStrap:
        return DeviceTypeDto.chestStrap;
      case DeviceType.headMounted:
        return DeviceTypeDto.headMounted;
      case DeviceType.smartDisplay:
        return DeviceTypeDto.smartDisplay;
    }
  }
}

/// Converts [DeviceTypeDto] to [DeviceType].
@sinceV1_0_0
@internal
extension DeviceTypeDtoToDomain on DeviceTypeDto {
  DeviceType toDomain() {
    switch (this) {
      case DeviceTypeDto.unknown:
        return DeviceType.unknown;
      case DeviceTypeDto.watch:
        return DeviceType.watch;
      case DeviceTypeDto.phone:
        return DeviceType.phone;
      case DeviceTypeDto.scale:
        return DeviceType.scale;
      case DeviceTypeDto.ring:
        return DeviceType.ring;
      case DeviceTypeDto.fitnessBand:
        return DeviceType.fitnessBand;
      case DeviceTypeDto.chestStrap:
        return DeviceType.chestStrap;
      case DeviceTypeDto.headMounted:
        return DeviceType.headMounted;
      case DeviceTypeDto.smartDisplay:
        return DeviceType.smartDisplay;
    }
  }
}
